src/app/public/src/modules/clipboard/clipboard.directive.ts

```diff
@@ -0,0 +1,80 @@
+// adapted from https://github.com/maxisam/ngx-clipboard/
+import {
+  Directive,
+  EventEmitter,
+  OnDestroy,
+  Output,
+  Renderer2,
+  ElementRef,
+  AfterViewInit,
+  ChangeDetectorRef,
+  Input
+} from '@angular/core';
+
+import { StacheClipboardService } from './clipboard.service';
+
+@Directive({
+  selector: '[stacheClipboard]'
+})
+export class StacheClipboardDirective implements AfterViewInit, OnDestroy {
+  // --------------------------------------------------------------
+  // NOTE: Only supported for Code-Block component as of July 2018
+  // --------------------------------------------------------------
+
+  private defaultTigger = 'clipboardButton';
+
+  @Input() public clipboardTrigger: string = this.defaultTigger;
+  @Input() public copySuccessText: string;
+  @Output() public cbOnSuccess: EventEmitter<any> = new EventEmitter<any>();
+  @Output() public cbOnError: EventEmitter<any> = new EventEmitter<any>();
+
+  private targetElm: any;
+  private button: HTMLButtonElement;
+
+  constructor(
+    private clipboardSrv: StacheClipboardService,
+    private cdr: ChangeDetectorRef,
+    private el: ElementRef,
+    private renderer: Renderer2
+  ) { }
+
+  public ngAfterViewInit() {
+    this.targetElm = this.el.nativeElement;
+    this.instantiateButton();
+    this.createEventListeners();
+    this.cdr.detectChanges();
+  }
+
+  public ngOnDestroy() {
+    this.clipboardSrv.destroy();
+  }
+
+  public onClick(): void {
+    const content = this.targetElm.innerText.trim();
+    this.handleResult(this.clipboardSrv.copyFromContent(content, this.renderer), content);
+  }
+
+  private handleResult(succeeded: Boolean, copiedContent: string | undefined): void {
+    if (succeeded) {
+      this.cbOnSuccess.emit({ isSuccess: true, content: copiedContent });
+    } else {
+      this.cbOnError.emit({ isSuccess: false });
+    }
+  }
+
+  private createEventListeners(): void {
+    if (this.button) {
+      this.renderer.listen(this.button, 'click', () => {
```

Comment:
This method reveals an implementation that's a bit muddy. You could instead add the directive to the trigger element (in this case, a button), and handle all the click events naturally. The target reference could also be passed in, so that the button would know what contents to copy.

Something like this?

```
<button
  stacheClipboard
  stacheClipboardTarget="stacheCodeBlock"
>
  Copy
</button>

<pre
  class="stache-code-output"
  stacheCodeBlock
>
  <code [ngClass]="getClassName()" [innerHTML]="output"></code>
</pre>
```